src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,910 @@
+        // describe a memory location within the object backing that ident.
+        // We conclude that ident is not assigned by this target and move on to
+        // the next target.
+      }
+    }
+    return false;
+  }
+
+  void apply()
+  {
+    std::size_t top_level_id = loop_nesting_graph.size();
+
+    // initialize the DECL ident -> loop id map with top level id
+    for(auto target = goto_function.body.instructions.begin();
+        target != goto_function.body.instructions.end();
+        target++)
+    {
+      if(target->is_decl())
+      {
+        ident_to_loop_id_map[target->decl_symbol().get_identifier()] =
+          top_level_id;
+      }
+    }
+
+    for(std::size_t loop_id = 0; loop_id < loop_nesting_graph.size(); loop_id++)
+    {
+      // Recursive traversal from each top level loop
+      // (top level loops do not have successors)
+      if(loop_nesting_graph.get_successors(loop_id).empty())
+      {
+        top_level_loops.push_back(loop_id);
+        apply_rec(loop_id);
+      }
+    }
+
+    // Compute set of locals at function level
+    top_level_local.insert(
+      goto_function.parameter_identifiers.begin(),
+      goto_function.parameter_identifiers.end());
+
+    for(const auto &entry : ident_to_loop_id_map)
+    {
+      if(entry.second == loop_info_map.size())
+      {
+        top_level_local.insert(entry.first);
+      }
+    }
+
+    // Compute subset of function locals that must be tracked in the function's
+    // write set. A local must be tracked if it is dirty or if it may be
+    // assigned by one of the top level loops.
+    for(const auto &ident : top_level_local)
+    {
+      if(dirty(ident))
+      {
+        top_level_tracked.insert(ident);
+      }
+      else
+      {
+        // Check if this ident is touched by one of the top level loops
+        for(const auto loop_id : top_level_loops)
+        {
+          if(is_assigned(ident, loop_info_map.at(loop_id).assigns))
+          {
+            top_level_tracked.insert(ident);
+          }
+        }
+      }
+    }
+  }
+
+  void apply_rec(const std::size_t loop_id)
+  {
+    // update the DECL ident -> loop id map on the way down
+    for(auto target : loop_nesting_graph[loop_id].instructions)
+    {
+      if(target->is_decl())
+      {
+        ident_to_loop_id_map[target->decl_symbol().get_identifier()] = loop_id;
+      }
+    }
+
+    // Process inner loops
+    for(const auto inner_loop_id : loop_nesting_graph.get_predecessors(loop_id))
+    {
+      apply_rec(inner_loop_id);
+    }
+
+    auto &loop = loop_nesting_graph[loop_id];
+
+    // Collect identifiers that are local to this loop only
+    // (excluding nested loop).
+    std::unordered_set<irep_idt> loop_local;
+    for(const auto &target : loop.instructions)
+    {
+      if(target->is_decl())
+      {
+        auto &ident = target->decl_symbol().get_identifier();
+        if(ident_to_loop_id_map[ident] == loop_id)
+          loop_local.insert(ident);
+      }
+    }
```

Comment:
This is a bit of a random point to make this note: haven't you called `tag_loop_instructions` before entering this function? So you could, as far as I understand, avoid a lot of the dance done here and just fetch the loop id from the instruction?! `dfcc_get_loop_id(target)` would directly tell you rather than having to go through `ident_to_loop_id_map`. Similar simplifications would apply in a lot of the above.

Reply:
Yes we could get rid of loop_id_map but then we depend on loop tagging being done before this step.

Reply:
got rid of loop_id_map but now the order of tagging and dfcc_loop_infot instances is dependent